### PR TITLE
Fix worker halting

### PR DESCRIPTION
**Pull request by gnodar01**
The problem:

From a debug log of running CP on a pipeline which encounters the issue of a hanging post-processing step, we have this as our final Progress before the hang:

```
[CP - DEBUG] cellprofiler_core.analysis._runner::interface: 👺 sending Progress - [(89, 'FinishedWaitingMeasurements'), (90, 'FinishedWaitingMeasurements'), (92, 'FinishedWaitingMeasurements'), (93, 'FinishedWaitingMeasurements'), (95, 'FinishedWaitingMeasurements'), (96, 'FinishedWaitingMeasurements'), (97, 'FinishedWaitingMeasurements'), (99, 'FinishedWaitingMeasurements'), (101, 'FinishedWaitingMeasurements')]

[CP - INFO] cellprofiler.gui.pipelinecontroller::analysis_event_handler: Progress, Counter({'Done': 231, 'FinishedWaitingMeasurements': 9})
```

I ran the pipeline on `240` image sets, so it wasn't the last few that hung, it was a bunch in the middle, almost sequentially.

Also from the logs, a successfully completed (progress `Done`) image set looks like this:

```
[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 job is: [240]
[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker sending MeasurementsReport for job [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::jobserver: Received measurements report for [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::queue_received_measurements: 👺 QUEUE OP: jobserver starting to put measurements in queue for interface, img_set_nums [240]; is not full
[CP - DEBUG] cellprofiler_core.analysis._runner::queue_received_measurements: 👺 QUEUE OP: jobserver done putting measurements in queue for interface, img_set_nums [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::jobserver: Acknowledged measurements report for [240]

[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker received MeasurementsReport ACK for job [240]

[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker finished with job [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::interface: 👺 deleting measurements for [240] as it has be consolidated
```

Whereas an unsuccessful one (stuck in `FinishedWaitingMeasurements` status), looks like this:

```
[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 job is: [101]
[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 worker sending MeasurementsReport for job [101]

[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 worker finished with job [101]
```

Once stuck, the worker blocks until analysis is terminated. So because we have 9 stuck above, those workers are not used for the remainder of jobs in the pipeline. That's pretty bad.

From the logs, the Worker does its job correctly. It gets a job, runs the pipeline, and sends back a `req` of type `MeasurementsReport`. It is the task of `Boundary.spin` to receive this over a socket, and send it to the upward queue for `interface` and `jobserver` to process. For the case of `MeasurementsReport`, it is the job of `interface` to consolidate it into the aggregate measurements, and then send back and `ACK` to the `Worker` so that it can stop blocking and go about its business.

The `interface` and `jobserver` never receive these hung jobs in their queues. In the log it appears `Boundary.spin` never even receives it.

### CODE REVIEW

**Comment by gnodar01** [comment]
The problem:

From a debug log of running CP on a pipeline which encounters the issue of a hanging post-processing step, we have this as our final Progress before the hang:

```
[CP - DEBUG] cellprofiler_core.analysis._runner::interface: 👺 sending Progress - [(89, 'FinishedWaitingMeasurements'), (90, 'FinishedWaitingMeasurements'), (92, 'FinishedWaitingMeasurements'), (93, 'FinishedWaitingMeasurements'), (95, 'FinishedWaitingMeasurements'), (96, 'FinishedWaitingMeasurements'), (97, 'FinishedWaitingMeasurements'), (99, 'FinishedWaitingMeasurements'), (101, 'FinishedWaitingMeasurements')]

[CP - INFO] cellprofiler.gui.pipelinecontroller::analysis_event_handler: Progress, Counter({'Done': 231, 'FinishedWaitingMeasurements': 9})
```

I ran the pipeline on `240` image sets, so it wasn't the last few that hung, it was a bunch in the middle, almost sequentially.

Also from the logs, a successfully completed (progress `Done`) image set looks like this:

```
[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 job is: [240]
[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker sending MeasurementsReport for job [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::jobserver: Received measurements report for [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::queue_received_measurements: 👺 QUEUE OP: jobserver starting to put measurements in queue for interface, img_set_nums [240]; is not full
[CP - DEBUG] cellprofiler_core.analysis._runner::queue_received_measurements: 👺 QUEUE OP: jobserver done putting measurements in queue for interface, img_set_nums [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::jobserver: Acknowledged measurements report for [240]

[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker received MeasurementsReport ACK for job [240]

[Worker 2 (76933)] cellprofiler_core.worker._worker::do_job: 👺 worker finished with job [240]

[CP - DEBUG] cellprofiler_core.analysis._runner::interface: 👺 deleting measurements for [240] as it has be consolidated
```

Whereas an unsuccessful one (stuck in `FinishedWaitingMeasurements` status), looks like this:

```
[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 job is: [101]
[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 worker sending MeasurementsReport for job [101]

[Worker 5 (77016)] cellprofiler_core.worker._worker::do_job: 👺 worker finished with job [101]
```

Once stuck, the worker blocks until analysis is terminated. So because we have 9 stuck above, those workers are not used for the remainder of jobs in the pipeline. That's pretty bad.

From the logs, the Worker does its job correctly. It gets a job, runs the pipeline, and sends back a `req` of type `MeasurementsReport`. It is the task of `Boundary.spin` to receive this over a socket, and send it to the upward queue for `interface` and `jobserver` to process. For the case of `MeasurementsReport`, it is the job of `interface` to consolidate it into the aggregate measurements, and then send back and `ACK` to the `Worker` so that it can stop blocking and go about its business.

The `interface` and `jobserver` never receive these hung jobs in their queues. In the log it appears `Boundary.spin` never even receives it.

**Comment by gnodar01** [comment]
One suspect is the following

In `Runner` we have:

```
        # We use a queue size of 10 because we keep measurements in memory (as
        # their HDF5 file contents) until they get merged into the full
        # measurements set.  If at some point, this size is too limiting, we
        # should have jobserver() call load_measurements_from_buffer() rather
        # than interface() doing so.  Currently, passing measurements in this
        # way seems like it might be buggy:
        # https://github.com/h5py/h5py/issues/244
        self.received_measurements_queue = queue.Queue(maxsize=10)
```

This queue is where `MeasurementReport` objects should go. It has `maxsize=10`.

In the `jobserver` we have:

```
elif isinstance(req, anarequest.MeasurementsReport):
    self.queue_received_measurements(req.image_set_numbers, req.buf)
    req.reply(anareply.Ack())
```

where `queue_received_measurements` is:

```
def queue_received_measurements(self, image_set_numbers, measurements):
        self.received_measurements_queue.put((image_set_numbers, measurements))
```

That will obviously encounter the `maxsize` limit on that queue. However this ends up not being (totally) responsible for the issue. If we were using `put_nowait` it would drop it, which would indeed be a problem and lead to a forever hanging `pos-processing` step, because the `Worker` never receives its `ACK`. However we use `put` instead. That will simply block until the queue is below the max capacitance of `10`. 

The `interface` gets from that queue, eventually, and then the `jobserver` continues.

While it is not *THE* problem, it may well be part of the problem. If the `jobserver` is blocked, its not processing other requests. This causes a backlog which is not relieved until `interface` reduces the queue size. That period in which `jobserver` is hung, allowing an accumulation which may exacerbate the underlying problem.

**Comment by gnodar01** [comment]
From [High Water Mark](https://zguide.zeromq.org/docs/chapter2/#High-Water-Marks) section of the ZMQ docs:

> When your socket reaches its HWM, it will either block or drop data depending on the socket type. PUB and ROUTER sockets will drop data if they reach their HWM, while other socket types will block.

In `Boundary.spin` we have a `ROUTER` socket:

```python
request_socket = self.zmq_context.socket(zmq.ROUTER)
request_socket.setsockopt(zmq.LINGER, 0)
request_port = request_socket.bind_to_random_port(
    self.zmq_address)
self.request_address = self.zmq_address + (":%d" % request_port)
```

In the `Worker`, we have a `work_socket` of type `REQ`, connected to that same address:

```python
self.work_socket = self.context.socket(zmq.REQ)
self.work_socket.connect(self.work_request_address)
```

The default highwater mark is `1000`, however also from the same section of the docs:

> Lastly, the HWMs are not exact; while you may get up to 1,000 messages by default, the real buffer size may be much lower (as little as half), due to the way libzmq implements its queues.

In any case, the `request_socket` in `Boundary.spin` is the thing that gets the `MeasurementsReport` from the worker, which is then placed in the upward queue for `interface` and `jobserver`. If it has reached the high water mark, it will drop `MeasurementsReport` message, because it is of type `ROUTER`. That means `jobserver` never gets it, and `interface` never consolidates it.

**Comment by gnodar01** [comment]
Another candidate cause, also due to the use of `ROUTER` is this:

> If you’re using ROUTER sockets, it’s remarkably easy to lose messages by accident, by sending malformed identity frames (or forgetting to send an identity frame). In general setting the ZMQ_ROUTER_MANDATORY option on ROUTER sockets is a good idea, but do also check the return code on every send call.

> 0MQ drops messages it can't route

It could be in some rare cases, that malformed identities are made and those messages never get delivered. 

I think this is less likely, because I have this line in `Boundary.spin`:

```python
req = Communicable.recv(s, routed=True)
if isinstance(req, MeasurementsReport):
    LOGGER.debug(f"👺 Boundary - received measurements report, {req.image_set_numbers}")
```

`Commmunicable.recv` shouldn't be affected by any malformed identities, because its the one that setups up future routing. Up to the point where `recv` is called, we're just using default zmq behavior, nothing custom.

I never receive the debug message for hung jobs, so they presumably never even make it to the `request_socket` (of type `ROUTER`) in the first place.

**Comment by gnodar01** [comment]
Modifying the high water mark, on either the send side or the receive side, did not seem to fix the problem.

In the end a somewhat "hacky" solution worked, which is simply to timeout and retry.

Ultimately a proper fix would likely require a larger refactor that includes thinking through and simplifying the concurrency system we currently have in place. That is a PR for another time.

File: src/subpackages/core/cellprofiler_core/worker/_worker.py [hunk 1]
```diff
@@ -280,26 +282,39 @@ def do_job(self, job):
...
+                        )
+                        # There might be an exception in this call, but it will be
+                        # handled elsewhere, and there's nothing we can do for it
+                        # here.
+                        current_pipeline.post_group(
+                            last_workspace, current_measurements.get_grouping_keys()
+                        )
+                        del last_workspace
+                    else:
+                        LOGGER.error("No workspace for from last image set, cannot run post group")
```
Comment (bethac07) [inline src/subpackages/core/cellprofiler_core/worker/_worker.py:299]: typo

**Review by bethac07 — APPROVED** [review]

**Pull request merged**